### PR TITLE
Logical Reduction

### PR DESCRIPTION
**Pull request by mrks**
This rule tries to simply logical expressions. Currently, it only looks at disjunctive chains, i.e., chains of ORs: `(a AND b) OR (a AND c) OR (a AND d)`. In this example, `a` can be extracted: `a AND (b OR c OR d)`. Furthermore, if the expression is used by a predicate, this rule extracts `a` into its own PredicateNode, which can then be independently pushed down or used for detecting joins. A good example can be seen in TPC-H query 19. Without this rule, `p_partkey = l_partkey` would not be available as a join predicate and the predicates on `l_shipmode` and `l_shipinstruct` could not be pulled below the join. Conjunctive chains, i.e., `(a OR b) AND (a OR c) -> a OR (b AND c)` are currently ignored, as they don't allow us to fully extract predicates. 

```
+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
| Benchmark | prev. iter/s  | runs  | new iter/s    | runs  | change | p-value (significant if <0.001) |
+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
| TPC-H 1   | 4.40900421143 | 265   | 4.36460161209 | 262   | -1%    |                          0.0000 |
| TPC-H 2   | 124.797393799 | 7488  | 113.566864014 | 6815  | -9%    |     (run time too short) 0.0000 |
| TPC-H 3   | 58.2987937927 | 3498  | 58.5161552429 | 3511  | +0%    |     (run time too short) 0.0000 |
| TPC-H 4   | 22.2972202301 | 1338  | 22.272518158  | 1337  | -0%    |     (run time too short) 0.2332 |
| TPC-H 5   | 48.0021324158 | 2881  | 48.3508872986 | 2902  | +1%    |     (run time too short) 0.0000 |
| TPC-H 6   | 120.377128601 | 7223  | 120.015518188 | 7201  | -0%    |     (run time too short) 0.0000 |
| TPC-H 7   | 13.1406164169 | 789   | 13.1597604752 | 790   | +0%    |     (run time too short) 0.2037 |
| TPC-H 8   | 82.6129455566 | 4957  | 81.509475708  | 4891  | -1%    |     (run time too short) 0.0000 |
| TPC-H 9   | 24.5970001221 | 1476  | 24.7648544312 | 1486  | +1%    |     (run time too short) 0.0000 |
| TPC-H 10  | 31.6822109222 | 1901  | 31.2141876221 | 1873  | -1%    |     (run time too short) 0.0000 |
| TPC-H 11  | 176.043670654 | 10000 | 176.888458252 | 10000 | +0%    |     (run time too short) 0.0000 |
| TPC-H 12  | 54.296585083  | 3258  | 53.4016571045 | 3205  | -2%    |     (run time too short) 0.0000 |
| TPC-H 13  | 27.8391914368 | 1671  | 26.7272014618 | 1604  | -4%    |     (run time too short) 0.0000 |
| TPC-H 14  | 138.862426758 | 8332  | 138.80480957  | 8329  | -0%    |     (run time too short) 0.0921 |
| TPC-H 15  | 61.6826438904 | 3701  | 62.2781295776 | 3737  | +1%    |     (run time too short) 0.0000 |
| TPC-H 16  | 69.757484436  | 4186  | 69.5473175049 | 4173  | -0%    |     (run time too short) 0.0000 |
| TPC-H 17  | 5.41747951508 | 326   | 4.41271018982 | 265   | -19%   |                          0.0000 |
| TPC-H 18  | 20.0286617279 | 1202  | 19.8063220978 | 1189  | -1%    |     (run time too short) 0.0000 |
| TPC-H 19  | 1.45210754871 | 88    | 33.9318809509 | 2036  | +2237% |     (run time too short) 0.0000 |
| TPC-H 20  | 3.27274465561 | 197   | 2.64355683327 | 159   | -19%   |                          0.0000 |
| TPC-H 21  | 1.54694449902 | 93    | 1.28587293625 | 78    | -17%   |                          0.0000 |
| TPC-H 22  | 115.943962097 | 6957  | 115.004966736 | 6901  | -1%    |     (run time too short) 0.0000 |
| average   |               |       |               |       | +98%   |                                 |
+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
```

### CODE REVIEW

**Comment by Bouncner** [comment]
Niiiiice!

Am Mo., 22. Okt. 2018 um 15:15 Uhr schrieb mrks <notifications@github.com>:

File: src/test/optimizer/strategy/logical_expression_reducer_rule_test.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    _d = equals_(LQPColumnReference(_table, ColumnID{3}), 3);
+    _e = equals_(LQPColumnReference(_table, ColumnID{4}), 4);
+  }
+
+  std::shared_ptr<AbstractLQPNode> _table;
+  std::shared_ptr<AbstractExpression> _a, _b, _c, _d, _e;
+};
+
+TEST_F(LogicalExpressionReducerRuleTest, CollectChainedAndExpressions) {
+  // a = 0 AND b = 1 AND (c = 2 OR d = 3)
```
Comment (timzimmermann) [inline src/test/optimizer/strategy/logical_expression_reducer_rule_test.cpp]:
```suggestion
  // a = 0 AND b = 1 AND (c = 2 OR d = 3) AND e = 4
```

**Comment by mrks** [comment]
Thanks.

Previously, we cheated and extracted the predicate needed for the join manually.

File: src/test/optimizer/strategy/logical_expression_reducer_rule_test.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+  {
+    ExpressionUnorderedSet chain;
+    Rule::_collect_chained_logical_expressions(expression, LogicalOperator::Or, chain);
+    EXPECT_EQ(chain, (ExpressionUnorderedSet{}));
+  }
+}
+
+TEST_F(LogicalExpressionReducerRuleTest, CollectChainedOrExpressions) {
+  // a = 0 OR b = 1 OR (c = 2 AND d = 3)
```
Comment (timzimmermann) [inline src/test/optimizer/strategy/logical_expression_reducer_rule_test.cpp]:
```suggestion
  // a = 0 OR b = 1 OR (c = 2 AND d = 3) OR e = 4
```

File: src/test/optimizer/strategy/logical_expression_reducer_rule_test.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+  {
+    ExpressionUnorderedSet chain;
+    Rule::_collect_chained_logical_expressions(expression, LogicalOperator::And, chain);
+    EXPECT_EQ(chain, (ExpressionUnorderedSet{}));
+  }
+}
+
+TEST_F(LogicalExpressionReducerRuleTest, RemoveFromChainedAndExpressions) {
+  // a = 0 AND b = 1 AND (c = 2 OR d = 3)
```
Comment (timzimmermann) [inline src/test/optimizer/strategy/logical_expression_reducer_rule_test.cpp]:
```suggestion
  // a = 0 AND b = 1 AND (c = 2 OR d = 3) AND e = 4
```

File: src/test/optimizer/strategy/logical_expression_reducer_rule_test.cpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    // No OR-chain in this example
+    auto expression_copy = expression->deep_copy();
+    Rule::_remove_expressions_from_chain(expression_copy, LogicalOperator::Or,
+                                         ExpressionUnorderedSet{_a, or_expression});
+    EXPECT_EQ(*expression_copy, *expression);
+  }
+}
+
+TEST_F(LogicalExpressionReducerRuleTest, RemoveFromChainedOrExpressions) {
+  // a = 0 OR b = 1 OR (c = 2 AND d = 3)
```
Comment (timzimmermann) [inline src/test/optimizer/strategy/logical_expression_reducer_rule_test.cpp]:
```suggestion
  // a = 0 OR b = 1 OR (c = 2 AND d = 3) OR e = 4
```

**Comment by mrks** [comment]
Ok, this "suggest changes" feature is a nice idea, but creating four commits for this is total BS. It could have waited if anything else happens in a minute or so and merge all those changes. I don't think I'll continue to apply those changes from the web interface.

**Comment by timzimmermann** [comment]
> Ok, this "suggest changes" feature is a nice idea, but creating four commits for this is total BS. It could have waited if anything else happens in a minute or so and merge all those changes. I don't think I'll continue to apply those changes from the web interface.

I totally agree, I already filed it with Github.

**Comment by mrzzzrm** [comment]
I very much like this idea.

Can we hold off from merging this until I had a look at it? It seems to me that the code is more complicated than it would need to be, but that's just first sight.

Also, the algorithm reads like you can rewrite more complicated expressions (`(b AND c AND d) OR (b AND c AND e) OR (x AND y AND whatnot AND b AND c)`), but the tests do not reflect that.

Also in the tests, you shouldn't need this empty table. For what you're doing, the "MockNode" was specifically designed.

**Comment by mrks** [comment]
> Also, the algorithm reads like you can rewrite more complicated expressions ((b AND c AND d) OR (b AND c AND e) OR (x AND y AND whatnot AND b AND c)), but the tests do not reflect that.

Mmh, never intended it, but I guess it works with two small changes. Done.

> Also in the tests, you shouldn't need this empty table. For what you're doing, the "MockNode" was specifically designed.

Ah, thanks. Fixed.

**Comment by mrks** [comment]
superseded by #1246
